lib/srv/desktop/rdp/rdpclient/src/client.rs

```diff
@@ -132,101 +139,125 @@ impl Client {
+                }
+            },
+            // One of the loops panicked. Abort the other and return the error.
+            Err(either) => match either {
+                futures_util::future::Either::Left((read_loop_panic, write_loop_handle)) => {
+                    write_loop_handle.abort();
+                    Err(read_loop_panic.into())
+                }
+                futures_util::future::Either::Right((write_loop_panic, read_loop_handle)) => {
+                    read_loop_handle.abort();
+                    Err(write_loop_panic.into())
+                }
+            },
+        }
     }
 
-    async fn run_rdp_loop_internal(mut self) -> Result<(), ClientError> {
-        if let Some(mut function_receiver) = self.function_receiver.take() {
+    fn run_read_loop(
+        cgo_handle: CgoHandle,
+        mut read_stream: RdpReadStream,
+        mut x224_processor: X224Processor,
+        write_requester: ClientHandle,
+    ) -> tokio::task::JoinHandle<Result<(), ClientError>> {
+        global::TOKIO_RT.spawn(async move {
             loop {
-                tokio::select! {
-                     res = self.rdp_stream.read_pdu() => {
-                        let (action, mut frame) = res?;
-                        match action {
-                            ironrdp_pdu::Action::X224 => {
-                                let result = self.process_x224_frame(&frame).await;
-                                self.process_active_stage_result(result).await?;
-                            },
-                            ironrdp_pdu::Action::FastPath => {
-                                unsafe {
-                                    handle_remote_fx_frame(self.cgo_handle, frame.as_mut_ptr(), frame.len() as u32);
-                                }
-                            },
-                        };
-                     }
-                     Some(data) = function_receiver.recv() => {
-                        trace!("Client received {:?}", data);
-                        match data {
-                            ClientFunction::WriteRdpKey(args) => {
-                                self.write_rdp_key(args).await?;
-                            },
-                            ClientFunction::WriteRdpPointer(args) => {
-                                self.write_rdp_pointer(args).await?;
-                            },
-                            ClientFunction::HandleResponsePdu(args) => {
-                                self.handle_response_pdu(args).await?;
-                            },
-                            ClientFunction::Stop => {
-                                return Ok(());
-                            }
-                        }
+                let (action, mut frame) = read_stream.read_pdu().await?;
+                match action {
+                    ironrdp_pdu::Action::FastPath => unsafe {
+                        handle_remote_fx_frame(cgo_handle, frame.as_mut_ptr(), frame.len() as u32);
```

Comment:
We still block on `handle_remote_fx_frame` without telling tokio about it, right?